hi-ml-multimodal/test_multimodal/image/model/test_model.py

```diff
@@ -156,3 +156,92 @@ def test_hubconf() -> None:
         if value_hub is None and value_himl is None:  # for example, class_logits
             continue
         assert torch.allclose(value_hub, value_himl)
+
+
+def test_load_biovil_weights() -> None:
+    """Test that the ``biovil`` option loads weights correctly."""
+    expected_model = get_biovil_resnet()
+
+    joint_feature_size = 128
+    biovil_model = ImageModel(img_model_type="resnet50",
+                              joint_feature_size=joint_feature_size,
+                              weights="biovil")
+
+    biovil_params = biovil_model.encoder.state_dict()
+    expected_params = expected_model.encoder.state_dict()
+
+    for (k, v), (exp_k, exp_v) in zip(
+        biovil_params.items(), expected_params.items()
+    ):
+        assert k == exp_k
+
+        # BN running means are different
+        if 'running' not in k:
```

Comment:
Why are they different?

Reply:
So this actually doesn't apply for the BioViL version as all parameters are over-written - I have removed that from the BioViL unit test. However, it does occur when initialising an ImageNet-pretrained ImageModel - I had noticed the same behaviour in the unchanged repo previously but didn't look into it. Specifically it is running/tracked parameters that are affected.

I've compared the ImageModel.encoder parameters with a model from "resnet50()" during various stages of initialisation and they diverge in get_encoder_output_dim(self.encoder) (line 90), specifically at line 215. Basically what I think is happening is that the parameters are identical until self.encoder is called on the random tensor in this function - the running/tracked params then get updated, but all weights and non-tracked params stay the same as no training has occurred.

I tried moving self.freeze_encoder/self.train() (lines 97/98) up above line 90, but that caused all sorts of other failures :(

Reply:
Thanks for the investigation. I'll open a PR that will hopefully fic this issue and then we can update this branch.

Reply:
PR opened:
- https://github.com/microsoft/hi-ml/pull/809

Reply:
I've merged the changes to this branch. Can you please see if we can safely remove those exceptional checks?